service/eventer/github/github.go

```diff
@@ -189,141 +128,72 @@ func (e *Eventer) NewDeploymentEvents() (<-chan eventerspec.DeploymentEvent, err
 		}
 	}
 
-	// From here on we watch for new deployment events and update the TPO
-	// accordingly.
+	var d eventerspec.DeploymentEvent
 	{
-		deploymentEventChannel, err := e.NewDeploymentEvents()
+		bytes, err := ioutil.ReadAll(res.Body)
 		if err != nil {
-			return microerror.Mask(err)
+			return eventerspec.DeploymentEvent{}, microerror.Mask(err)
 		}
 
-		for d := range deploymentEventChannel {
-			tpo, err := e.getTPO()
-			if err != nil {
-				return microerror.Mask(err)
-			}
-
-			newProject := draughtsmantprspec.Project{
-				Name: d.Name,
-				Ref:  d.Sha,
-			}
-
-			tpo.Spec.Projects = ensureProject(tpo.Spec.Projects, newProject)
+		var deployments []deployment
+		if err := json.Unmarshal(bytes, &deployments); err != nil {
+			return eventerspec.DeploymentEvent{}, microerror.Mask(err)
+		}
 
-			err = e.ensureTPO(tpo)
+		for i, depl := range deployments {
+			deploymentStatuses, err := e.fetchDeploymentStatus(project, depl)
 			if err != nil {
-				return microerror.Mask(err)
+				return eventerspec.DeploymentEvent{}, microerror.Mask(err)
 			}
-		}
-	}
-
-	return nil
-}
 
-func (e *Eventer) getTPO() (draughtsmantpr.CustomObject, error) {
-	endpoint := e.draughtsmanTPR.Endpoint(DefaultNamespace) + "/" + TPOName
-
-	b, err := e.k8sClient.Core().RESTClient().Get().AbsPath(endpoint).DoRaw()
-	if apierrors.IsNotFound(err) {
-		return draughtsmantpr.CustomObject{}, microerror.Mask(notFoundError)
-	} else if err != nil {
-		return draughtsmantpr.CustomObject{}, microerror.Mask(err)
-	}
-
-	var tpo draughtsmantpr.CustomObject
-	err = json.Unmarshal(b, &tpo)
-	if err != nil {
-		return draughtsmantpr.CustomObject{}, microerror.Mask(err)
-	}
-
-	return tpo, nil
-}
-
-// TODO create a TPO service in a separate PR to ease testing.
-func (e *Eventer) ensureTPO(tpo draughtsmantpr.CustomObject) error {
-	endpoint := e.draughtsmanTPR.Endpoint(DefaultNamespace) + "/" + TPOName
-	_, err := e.k8sClient.Core().RESTClient().Post().Body(tpo).AbsPath(endpoint).DoRaw()
-	if apierrors.IsNotFound(err) {
-		return microerror.Mask(notFoundError)
-	} else if apierrors.IsAlreadyExists(err) {
-		_, err := e.k8sClient.Core().RESTClient().Put().Body(tpo).AbsPath(endpoint).DoRaw()
-		if apierrors.IsNotFound(err) {
-			return microerror.Mask(notFoundError)
-		} else if err != nil {
-			return microerror.Mask(err)
+			deployments[i].Statuses = deploymentStatuses
 		}
-	} else if err != nil {
-		return microerror.Mask(err)
-	}
 
-	return nil
-}
+		deployments = e.filterDeploymentsByStatus(deployments)
 
-func ensureProject(projects []draughtsmantprspec.Project, project draughtsmantprspec.Project) []draughtsmantprspec.Project {
-	var updated bool
-	for _, p := range projects {
-		if p.Name != project.Name {
-			continue
+		if len(deployments) == 0 {
+			return eventerspec.DeploymentEvent{}, microerror.Mask(notFoundError)
```

Comment:
I don't follow here - why would this be zero? Why does that mean not found?

Reply:
My thinking is that we could just not have any deployments

Reply:
Don't know. Just being defensive. Would you not return an error? I think I am also fine with this.

Reply:
I think I wouldn't. I feel like a list being empty is different to the list not being found?

If you fix that up, I'll approve <3

Reply:
The thing here is that this is about fetching the latest deployment event. The caller needs to get one or do at least nothing in case there is none. To control this process I chose returning the error. It is totally fine for the caller to not do anything in case the error appears. The alternative would be to return an empty deployment event but this feels more odd. WDYT?